File: frontend/src/api/custom-fields.ts
import { request } from "@/api/client";
import type { CustomField } from "@/api/types";

export function listCustomFields(): Promise<CustomField[]> {
  return request<CustomField[]>("/api/custom-fields");
}

export function createCustomField(input: {
  field_type: string;
  name: string;
  group_name?: string | null;
  options?: string[] | null;
  is_required?: boolean;
}): Promise<CustomField> {
  return request<CustomField>("/api/custom-fields", {
    method: "POST",
    body: JSON.stringify(input),
  });
}

export function deleteCustomField(fieldId: string): Promise<void> {
  return request(`/api/custom-fields/${fieldId}`, { method: "DELETE" });
}

export function getCustomValues(personId: string): Promise<Record<string, unknown>> {
  return request<Record<string, unknown>>(`/api/persons/${personId}/custom-values`);
}

export function setCustomValues(
  personId: string,
  values: Record<string, unknown>,
): Promise<Record<string, unknown>> {
  return request<Record<string, unknown>>(`/api/persons/${personId}/custom-values`, {
    method: "PUT",
    body: JSON.stringify({ values }),
  });
}

File: backend/app/api/custom_fields.py
"""自定义字段 API。"""

from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends, Response
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.core.response import success_response
from app.schemas.p1 import (
    CustomFieldCreate,
    CustomFieldRead,
    CustomFieldUpdate,
    CustomFieldValueUpdate,
)
from app.services.custom_field_service import CustomFieldService

router = APIRouter()


@router.get("/custom-fields")
def list_custom_fields(db: Session = Depends(get_db)) -> dict:
    fields = CustomFieldService(db).list_fields()
    return success_response([CustomFieldRead.model_validate(item) for item in fields])


@router.post("/custom-fields")
def create_custom_field(data: CustomFieldCreate, db: Session = Depends(get_db)) -> dict:
    field = CustomFieldService(db).create_field(data)
    return success_response(CustomFieldRead.model_validate(field))


@router.patch("/custom-fields/{field_id}")
def update_custom_field(
    field_id: uuid.UUID, data: CustomFieldUpdate, db: Session = Depends(get_db)
) -> dict:
    field = CustomFieldService(db).update_field(field_id, data)
    return success_response(CustomFieldRead.model_validate(field))


@router.delete("/custom-fields/{field_id}")
def delete_custom_field(field_id: uuid.UUID, db: Session = Depends(get_db)) -> Response:
    CustomFieldService(db).delete_field(field_id)
    return Response(status_code=204)


@router.get("/persons/{person_id}/custom-values")
def get_custom_values(person_id: uuid.UUID, db: Session = Depends(get_db)) -> dict:
    return success_response(CustomFieldService(db).get_values(person_id))


@router.put("/persons/{person_id}/custom-values")
def set_custom_values(
    person_id: uuid.UUID,
    data: CustomFieldValueUpdate,
    db: Session = Depends(get_db),
) -> dict:
    values = CustomFieldService(db).set_values(person_id, data.values)
    return success_response(values)
